### PR TITLE
[HttpClient] allow to use unix socket when using amphp http client

### PR DESCRIPTION
**Pull request by praswicaksono**
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | -
| License       | MIT

Allow HttpClient to interact with unix socket when using amphp/http-client v5

```php
$client = new AmpHttpClient(clientConfigurator: function() {
    $connector = new StaticSocketConnector("unix:///var/run/docker.sock", socketConnector());

    return (new HttpClientBuilder)
        ->usingPool(new UnlimitedConnectionPool(new DefaultConnectionFactory($connector)))
        ->build();
});

$client = $client->withOptions(['base_uri' => 'http://docker']);

$response = $client->request('GET', '/info');
```

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Thanks for the PR. The change looks legit, but I'm wondering: can you try using option `bindto` to configure the socket instead of going with a custom configurator?

**Comment by praswicaksono** [comment]
@nicolas-grekas  Yes it works when using `bindto` and I updated test case also by using `bindto`

File: src/Symfony/Component/HttpClient/Internal/AmpListenerV5.php [hunk 1]
```diff
@@ -20,6 +20,10 @@
 use Amp\Http\Client\Response;
 use Symfony\Component\HttpClient\Exception\TransportException;
 
+if (\PHP_VERSION_ID < 80400 || !class_exists('Amp\Socket\UnixAddress')) {
+    throw new \LogicException('Using "Symfony\Component\HttpClient\AmpHttpClient" with amphp/http-client >= 5 requires PHP >= 8.4.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Internal/AmpListenerV5.php:24]: this "if" should be removed, it's already checked in AmpHttpClient

File: src/Symfony/Component/HttpClient/CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+7.3
+--
+
+* Allow `AmphpHttpClient` with amphp/http-client v5 to use unix socket
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/CHANGELOG.md:7]: this should be reverted: the change is a bugfix, not a new feature (the PR should target branch

File: src/Symfony/Component/HttpClient/Internal/AmpListenerV5.php [hunk 3]
```diff
@@ -66,14 +70,20 @@ public function connectionAcquired(Request $request, Connection $connection, int
 
     public function requestHeaderStart(Request $request, Stream $stream): void
     {
-        $host = $stream->getRemoteAddress()->getAddress();
+        $unixSocketClass = 'Amp\Socket\UnixAddress';
+        if ($stream->getRemoteAddress() instanceof $unixSocketClass) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Internal/AmpListenerV5.php:74]: lets inline the class, no need for the variable

File: src/Symfony/Component/HttpClient/Tests/AmpHttpClientTest.php [hunk 4]
```diff
@@ -24,6 +30,27 @@ protected function getHttpClient(string $testCase): HttpClientInterface
         return new AmpHttpClient(['verify_peer' => false, 'verify_host' => false, 'timeout' => 5]);
     }
 
+    public function testUnixSocket()
+    {
+        if (\PHP_VERSION_ID < 80400) {
+            $this->markTestSkipped('Unix sockets only support PHP >= 8.4.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Tests/AmpHttpClientTest.php:36]: this should be replaced by an `@requires PHP 8.4` annotation

File: src/Symfony/Component/HttpClient/Tests/AmpHttpClientTest.php [hunk 5]
```diff
@@ -24,6 +30,27 @@ protected function getHttpClient(string $testCase): HttpClientInterface
...
+            $this->markTestSkipped('Unix sockets are not supported on Windows.');
+        }
+
+        $client = new AmpHttpClient(['bindto' => '/var/run/docker.sock']);
+
+        $client = $client->withOptions(['base_uri' => 'http://docker']);
+
+        $response = $client->request('GET', '/info');
+
+        $this->assertSame(200, $response->getStatusCode());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Tests/AmpHttpClientTest.php:49]: this test case is going to work only if a docker daemon is running, isn't it?
I guess we need to skip the test case if the `docker.sock` file does not exist (and thus remove the windows check, which would be redundant)

**Review by nicolas-grekas — COMMENTED** [review]
What about amp/http-client v4?
The PR should target branch 7.2 as it's a bugfix also. Can you please rebase+retarget?

**Comment by praswicaksono** [comment]
> What about amp/http-client v4? The PR should target branch 7.2 as it's a bugfix also. Can you please rebase+retarget?
I will close this PR and redone it in 7.2

**Pull request closed without merging**